go/vt/vttablet/tabletmanager/rpc_external_reparent.go

```diff
@@ -193,51 +200,117 @@ func (agent *ActionAgent) finalizeTabletExternallyReparented(ctx context.Context
+	// here anymore. The lock was only to ensure that the global shard record
+	// didn't get modified between the time when we read it and the time when we
+	// write it back. Now we use an update loop pattern to do that instead.
+
+	// We also used to do this in parallel with RefreshState on the old master
+	// we don't do that any more. we want to update the shard record first
+	// and only then attempt to refresh the old master because it is possible
+	// that the old master is unreachable
+	event.DispatchUpdate(ev, "updating global shard record")
+	log.Infof("finalizeTabletExternallyReparented: updating global shard record if needed")
+	_, err = agent.TopoServer.UpdateShardFields(ctx, masterTablet.Keyspace, masterTablet.Shard, func(currentSi *topo.ShardInfo) error {
+		if topoproto.TabletAliasEqual(currentSi.MasterAlias, masterTablet.Alias) {
+			// It is correct to return this error here, UpdateShardFields will ignore it
+			return topo.NewError(topo.NoUpdateNeeded, masterTablet.Alias.String())
 		}
-	}()
+		if !topoproto.TabletAliasEqual(currentSi.MasterAlias, oldMasterAlias) {
+			log.Warningf("old master alias (%v) not found in the global Shard record i.e. it has changed in the meantime."+
+				" We're not overwriting the value with the new master (%v) because the current value is probably newer."+
+				" (initial Shard record = %#v, current Shard record = %#v)",
+				oldMasterAlias, masterTablet.Alias, si, currentSi)
+			// It is correct to return this error here, UpdateShardFields will ignore it
+			return topo.NewError(topo.NoUpdateNeeded, oldMasterAlias.String())
+		}
+		currentSi.MasterAlias = masterTablet.Alias
+		return nil
+	})
+	if err != nil {
+		errs.RecordError(err)
+	}
+
+	if errs.HasErrors() {
+		return errs.Error()
+	}
+
 	if !topoproto.TabletAliasIsZero(oldMasterAlias) {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
+		// Tell the old master to re-read its tablet record and change its state.
+		// We don't need to put error into errs if this fails, but we need to wait
+		// for it to make sure that old master tablet is not stuck in the MASTER
+		// state.
+		tmc := tmclient.NewTabletManagerClient()
+		if err := tmc.RefreshState(ctx, oldMasterTablet); err != nil {
+			log.Warningf("Error calling RefreshState on old master %v: %v", topoproto.TabletAliasString(oldMasterTablet.Alias), err)
+		}
+	}
 
-			// Tell the old master to re-read its tablet record and change its state.
-			// We don't need to put error into errs if this fails, but we need to wait
-			// for it to make sure that old master tablet is not stuck in the MASTER
-			// state.
-			tmc := tmclient.NewTabletManagerClient()
-			if err := tmc.RefreshState(ctx, oldMasterTablet); err != nil {
-				log.Warningf("Error calling RefreshState on old master %v: %v", topoproto.TabletAliasString(oldMasterTablet.Alias), err)
-			}
-		}()
+	tabletMap, err := agent.TopoServer.GetTabletMapForShard(ctx, masterTablet.Keyspace, masterTablet.Shard)
```

Comment:
Add a comment here to introduce this section: "Look for other false masters and clean them up on a best-effort basis."